Implement client authentication for backend TLS

tsaarni opened the pull request:
* Added new parameter to configuration file to configure a name of a secret
  containing client certificate and private key.

* Configure Envoy to authenticate itself towards upstream cluster with
  the client certificate when upstream connection uses TLS.

Closes #2338 

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
I've documented a manual test procedure here https://gist.github.com/tsaarni/245584e8b4e40b9aa5a9924c7d13eadf to validate following on a target environment:

* successful connection with valid client certificate
* rotate client certificate
* unsuccessful connection with invalid client certificate
* unsuccessful connection without client certificate

To run the procedure fully requires Kind on linux (I think...  the wireshark bits to be able to see TLS handshake)

youngnick commented on cmd/contour/servecontext.go:
```diff
@@ -205,37 +205,49 @@ type TLSConfig struct {
-// FallbackCertificate defines the namespace/name of the Kubernetes secret to
-// use as fallback when a non-SNI request is received.
-type FallbackCertificate struct {
+// NamespacedName defines the namespace/name of the Kubernetes resource referred
```
I wasn't sure at first why you weren't using `types.NamespacedName` here, but now I see it's used for the yaml value and for validation. Maybe something in this comment explaining that?

tsaarni replied:
I was thinking the same when renaming/reusing `FallbackCertificate` type for the new certificate:  Wouldn't Kubernetes `types.NamespacedName` have worked as well? The Kubernetes type does not have YAML tags, but isn't the default "untagged" behavior the same as the tags here? Or would there be some differences?

I was then reasoning that `FallbackCertificate` was defined likely because we did not want to depend on external type for our configuration file. In theory `types.NamespacedName` could change at any time, e.g. new field could be introduced which would automatically change parsing of our configuration file as well.  

@stevesloka: I wonder if this was the motivation for defining `type FallbackCertificate` and not using K8s `types.NamespacedName`? If yes, I'll write that down in the comment.

tsaarni replied:
I added a comment about the type being used for YAML parsing with the above assumption, hope the assumption was correct!

youngnick reviewed (approved):
LGTM with one small suggestion. I'll add @jpeach as well though for his review.

skriss commented on internal/contour/secret.go:
```diff
@@ -107,6 +107,10 @@ func (v *secretVisitor) visit(vertex dag.Vertex) {
 			v.addSecret(svh.FallbackCertificate)
 		}
+	case *dag.Cluster:
+		if svh.ClientCertificate != nil {
```
nit: let's rename `svh` now that it's not only for a `SecureVirtualHost` -- `obj` or something similarly generic.

tsaarni replied:
Fixed!

skriss commented on cmd/contour/serve.go:
```diff
@@ -247,11 +258,13 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 					FieldLogger: log.WithField("context", "IngressProcessor"),
 				},
```
should we be supporting this for `Ingress` too?

tsaarni replied:
I missed mentioning this in the design doc, but I've understood that there is no intention to support backend TLS for `Ingress`.  So I believe upstream TLS code is missing completely there.  This is mentioned in older design doc [design/tls-backend-verification.md](https://github.com/projectcontour/contour/blob/main/design/tls-backend-verification.md).

tsaarni replied:
Yes sorry, it is the server authentication that is missing: TLS as itself is supported using annotation `projectcontour.io/upstream-protocol.tls` on service like you say, but server authentication can be only expressed with `HTTPProxy` `spec.routes.services[].validation`. 

We can still add client authentication regardless of not authenticating the server, though it is unusual to have it this way.

tsaarni replied:
I've now added support for client auth for `Ingress` too.

skriss reviewed (commented):
this looks great to me @tsaarni, just a couple things.

skriss commented:
Unfortunately you'll need a rebase as well, there are some merge conflicts.

tsaarni commented:
Thanks for reviews so far! 

I've rebased the PR.

Note: I've also added draft of integration test which was not previously included in the PR, to automate the manual test procedure.  The tests will be skipped by default, and I can gladly also remove them for now if you wish!  The test depends on echoserver with my own changes to ingress-controller-conformance echoserver which add TLS support. I sent a PR but it might not work out at the end, so it may be that another approach should be taken at the end.

jpeach commented:
xref https://github.com/kubernetes-sigs/ingress-controller-conformance/pull/71

jpeach reviewed (changes requested):
I haven't finished reviewing yet, but here's the first set of feedback.

tsaarni reviewed (commented):
I would like to double check and confirm with you that we agree following is wanted behavior: 

When Envoy client certificate is configured, but the Secret does not exist, the clusters will not be created at all.  I believe it mimics how FallbackCertificate was implemented.   Is this your expectation here too? This makes the configuration error very visible, and error can be seen in all `HTTPProxy` statuses but it will be quite easy to break everything by removing just a single Secret.  

Alternative could be not configure client certificate to the upstream clusters while Secret is missing (or when it is broken).

tsaarni commented on internal/dag/httpproxy_processor.go:
```diff
@@ -495,6 +499,15 @@ func (p *HTTPProxyProcessor) computeRoutes(
+					return nil
+				}
+			}
+
```
I've added test cases for `IngressProcessor` and `ExtensionServiceProcessor` and also added tests where Secret is removed.

tsaarni commented on site/docs/main/httpproxy.md:
````diff
@@ -349,6 +349,8 @@ spec:
 ```
 
+Contour can be configured with a `namespace/name` in the [Contour configuration file][11] of a Kubernetes secret which Envoy will present as client certificate when establishing TLS connection to the backend.
+
````
I've added new section and explained these.

tsaarni commented on cmd/contour/servecontext_test.go:
```diff
@@ -290,6 +290,72 @@ func TestFallbackCertificateParams(t *testing.T) {
+
+			if diff := cmp.Diff(tc.want, got); diff != "" {
+				t.Fatal(diff)
+			}
```
Fixed.

tsaarni commented on internal/dag/extension_processor.go:
```diff
@@ -27,13 +27,21 @@ import (
 type ExtensionServiceProcessor struct {
 	logrus.FieldLogger
+
+	source *KubernetesCache
```
It is not needed, my bad!

tsaarni commented on internal/dag/extension_processor.go:
```diff
@@ -27,13 +27,21 @@ import (
+	p.source = source
+
+	for _, e := range source.extensions {
+		if ext := p.buildExtensionService(source, e); ext != nil {
```
Fixed.

tsaarni commented on site/docs/main/configuration.md:
```diff
@@ -54,6 +55,15 @@ Contour should provision TLS hosts.
 {: class="table thead-dark table-bordered"}
 <br>
 
+### Client Certificate
```
Fixed.

tsaarni commented on site/docs/main/configuration.md:
```diff
@@ -54,6 +55,15 @@ Contour should provision TLS hosts.
+| Field Name | Type| Default  | Description |
+|------------|-----|----------|-------------|
+| name       | string | `""` | This field specifies the name of the Kubernetes secret to use as the client certificate and private key when establishing TLS connection to the backend service. |
+| namespace  | string | `""` | This field specifies the namespace of the Kubernetes secret to use as the client certificate and private key when establishing TLS connection to the backend service. |
```
Fixed.

youngnick commented:
I feel like it's okay for this to behave the same as `fallbackCertificate`, personally. @jpeach @skriss @stevesloka ?

jpeach commented:
> I feel like it's okay for this to behave the same as `fallbackCertificate`, personally. @jpeach @skriss @stevesloka ?

Yeh, I think that this is a problem to handle in Contour in a more general way ...

jpeach commented on _integration/testsuite/httpproxy/013-backend-tls.yaml:
```diff
@@ -0,0 +1,262 @@
+$apply: delete
+
+---
+
```
Each fragment gets its own timeout, so the test ends up being more deterministic if you test what you expect in each stage in a separate fragment. Even with this, I think it's possible for the check to spuriously succeed since deleted objects can be visible through the API for a short time.

jpeach reviewed (approved):
Nice work @tsaarni !

jpeach commented:
@skriss LMK if you want to make another review pass. If you are OK, I'll merge this tomorrow.

tsaarni commented:
> Nice work @tsaarni !

Thanks, and thanks for great review comments!

BTW: I still did a small change & force push, which only contains typo fix for the doc and integration test improvement for certificate rotation.

skriss reviewed (approved):
LGTM!

jpeach commented:
@tsaarni can you please rebase?

tsaarni commented:
PR rebased!

moderation commented:
@tsaarni does the PR close https://github.com/projectcontour/contour/issues/2885? I don't _think_ so but wanted to check.

tsaarni commented:
> @tsaarni does the PR close #2885? I don't _think_ so but wanted to check.

No, this one was unrelated to #2885.
